client/src/helpers/storeSearch.ts

```diff
@@ -0,0 +1,340 @@
+    ed[i] = new Array(n + 1).fill(0);
+    ed[i][0] = i;
+  }
+
+  for (let i = 1; i < m + 1; i++) {
+    for (let j = 1; j < n + 1; j++) {
+      if (s1[i - 1] === s2[j - 1]) {
+        ed[i][j] = ed[i - 1][j - 1];
+      } else {
+        ed[i][j] = Math.min(ed[i - 1][j], ed[i][j - 1], ed[i - 1][j - 1]) + 1;
+      }
+    }
+  }
+
+  const min_j = ed[m].indexOf(Math.min(...ed[m]));
+  const min_ed = ed[m][min_j];
+  return [min_ed, n - min_j, min_j - m + min_ed];
+};
+
+const getNgrams = (s: string, n: number = DEFAULT_NGRAM_LENGTH) => {
+  const ngrams = [];
+  for (let i = n; i <= s.length; i++) {
+    ngrams.push(s.slice(i - n, i));
+  }
+  return ngrams;
+};
+
+const buildNgramIndex = (stores: Store[], n: number = DEFAULT_NGRAM_LENGTH) => {
+  const index = new Map(); // performs better than Object type for frequent additions
+  stores.forEach((store, id) => {
+    if (store.preprocessedName) {
+      const ngrams = getNgrams(store.preprocessedName);
+      ngrams.forEach((ngram) => {
+        if (!index.has(ngram)) {
+          index.set(ngram, []);
+        }
+        index.get(ngram).push(id);
+      });
+    }
+  });
+  return index;
+};
+
+/* Comparator function that defines the sort order of an array of stores. */
+const storeCompareFn = (
+  a: { storeId: number; dist: number[] },
+  b: { storeId: number; dist: number[] }
+) => {
+  return (
+    a.dist[0] - b.dist[0] || a.dist[1] - b.dist[1] || a.dist[2] - b.dist[2]
+  );
+};
+
+/**
+ * Initialize the variables needed for the Single-Pass Search algorithm, with heap.
+ *
+ * @param stores
+ * @param preprocessedQuery
+ * @returns lists - the list of ngram inverted list indices, corresponding to the
+ *                  ngrams of the preprocessed query.
+ * @returns listIndices - the index of the next store for each list.
+ * @returns candidateStores - a heap to hold the next store IDs from each list.
+ * @returns topKStores - a heap to hold the top k stores.
+ */
+const initializeSPS = (stores: Store[], preprocessedQuery: string) => {
+  // Get ngram inverted list indices
+  const ngramIndex = buildNgramIndex(stores);
+  const queryNgrams: string[] = getNgrams(preprocessedQuery);
+  const lists = queryNgrams.map((ngram) =>
+    ngramIndex.has(ngram) ? ngramIndex.get(ngram) : []
+  );
+
+  const topKStores = new Heap<{ storeId: number; dist: number[] }>(
+    storeCompareFn
+  );
+  const listIndices = new Array(lists.length).fill(0);
+  const candidateStores = new Heap<number[]>((a, b) => b[0] - a[0]);
+  listIndices.forEach((idx, i) => {
+    if (idx < lists[i].length) {
+      candidateStores.insert([lists[i][idx], i]);
+      listIndices[i] = 1;
+    }
+  });
+  return { lists, listIndices, candidateStores, topKStores };
+};
+
+const calculateMinFreq = (
+  nQueries: number,
+  maxEdits: number,
+  n: number = 3
+) => {
+  return Math.max(0, nQueries - n * maxEdits);
+};
+
+const popElementsByValue = (heap: Heap<any>, poppedLists: number[]) => {
```

Comment:
WHy Heap<any> ?
```suggestion
const popElementsByValue = (heap: Heap<??>, poppedLists: number[]) => {
```

Reply:
this change seems to cause errors? e.g. `'>' expected.`

Reply:
Using a heap allows us to traverse the ngram inverted list indices by order of store index. Compared to doing round robin over the lists,
> Advantages: 1) No need to maintain the set of candidate IDs, and 2) More opportunities to skip stores.

Will add more details in the explanation in the design doc

Reply:
Sorry i meant why <any> it should have a defined type

Reply:
Oh, I'm just using T as a generic type variable https://www.typescriptlang.org/docs/handbook/generics.html 
So that it defines the type of each element in the heap, and is more precise than using `any`

Reply:
okay thanks!